### PR TITLE
chore: Reorganize feature logging in Go for better extensibility

### PR DESCRIPTION
**Pull request by pyalex**
Signed-off-by: pyalex <moskalenko.alexey@gmail.com>

**What this PR does / why we need it**:

This is mostly refactoring PR. This refactoring is needed before integrating Python API for feature logging in offline store.
Improvements:
1. Splitting LoggingService (single instance) with Logger (one per feature service) - allows to cache logger (per feature service) instead of creating new one on each request.
2. Decoupling schema generation from logger - improves testing.
3. Decoupling buffer operations - improves testing.
4. Simplifying sink logic - there will be one prod and one test sink in foreseeable future, it's better to keep things simple - improves testing.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -105,7 +105,7 @@ coverage.xml
 .hypothesis/
 .pytest_cache/
 infra/scripts/*.conf
-go/cmd/server/logging/feature_repo/data/
+go/intenal/test/feature_repo/data/
```
Comment (kevjumba) [inline .gitignore:108]: internal

File: go/cmd/server/server_test.go [hunk 2]
```diff
@@ -232,26 +234,23 @@ func TestGetOnlineFeaturesSqliteWithLogging(t *testing.T) {
...
-					assert.Equal(t, featureVal.Val, expectedLogValues[name].Val[idx].Val)
+					assert.Equal(t, featureVal.Val, actualValues[name].Val[idx].Val)
 				}
 			}
 
 		}
 	}
-	err = test.CleanUpFile(logPath)
-	assert.Nil(t, err)
 }
```
Comment (kevjumba) [inline go/cmd/server/server_test.go:254]: Do we no longer need to clean up the generated log file?

Comment (pyalex) [reply go/cmd/server/server_test.go:254]: Yep, it's written to temp directory, which will be cleaned up by test framework

File: go/internal/feast/logging/filelogsink.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// and the parquet files will be cleaned up after the test is run.
+func NewFileLogSink(path string) (*FileLogSink, error) {
+	store := FileLogSink{}
+	var absPath string
+	var err error
+	// TODO(kevjumba) remove this default catch.
+	if path != "" {
+		absPath, err = filepath.Abs(path)
+	} else {
+		return nil, errors.New("need path for file log storage")
```
Comment (kevjumba) [inline go/internal/feast/logging/filelogsink.go:33]: file log sink?

File: go/internal/feast/logging/logger.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+			}
+			lErr = errors.WithStack(rErr)
+		}
+	}()
+	for {
+		shouldStop := false
+
+		select {
+		case <-l.signalCh:
+			l.buffer.writeBatch(l.sink)
```
Comment (kevjumba) [inline go/internal/feast/logging/logger.go:136]: I think this should also have error handling for writing batch.

File: go/internal/feast/logging/logger.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+			l.sink.Flush()
+			shouldStop = true
+		case <-time.After(l.config.WriteInterval):
```
Comment (kevjumba) [inline go/internal/feast/logging/logger.go:139]: Maybe add some comments here that specify that differentiates flushing from writing.

File: go/internal/feast/logging/featureserviceschema.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+				if _, ok := joinKeysSet[joinKey]; !ok {
+					joinKeys = append(joinKeys, joinKey)
+				}
+
+				joinKeysSet[joinKey] = nil
+				entityJoinKeyToType[joinKey] = entity.ValueType
+			}
+		} else if odFv, ok := odFvMap[featureViewName]; ok {
+			for _, f := range featureProjection.Features {
+				// TODO(kevjumba) check in test here.
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/feast/logging/featureserviceschema.go]: Removed

File: go/internal/feast/logging/filelogsink.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/apache/arrow/go/v8/arrow"
+	"github.com/google/uuid"
+
+	"github.com/apache/arrow/go/v8/arrow/array"
+	"github.com/apache/arrow/go/v8/parquet"
+	"github.com/apache/arrow/go/v8/parquet/pqarrow"
+)
+
+type FileLogSink struct {
+	// Feast project name
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/feast/logging/filelogsink.go]: removed

File: go/internal/feast/logging/filelogsink.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	path string
+}
+
+// FileLogSink is currently only used for testing. It will be instantiated during go unit tests to log to file
+// and the parquet files will be cleaned up after the test is run.
+func NewFileLogSink(path string) (*FileLogSink, error) {
+	store := FileLogSink{}
+	var absPath string
+	var err error
+	// TODO(kevjumba) remove this default catch.
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/feast/logging/filelogsink.go]: removed

File: go/internal/feast/logging/logger_test.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+func (s *DummySink) Write(rec arrow.Record) error {
+	return nil
+}
+
+func (s *DummySink) Flush() error {
+	return nil
+}
+
+func TestLoggingChannelTimeout(t *testing.T) {
+	// Pregenerated using `feast init`.
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/feast/logging/logger_test.go]: removed

File: go/internal/test/go_integration_test_utils.go [hunk 10]
```diff
@@ -175,22 +176,14 @@ func CleanUpInitializedRepo(basePath string) {
...
-}
-
 func GetProtoFromRecord(rec array.Record) (map[string]*types.RepeatedValue, error) {
 	r := make(map[string]*types.RepeatedValue)
 	schema := rec.Schema()
 	for idx, column := range rec.Columns() {
 		field := schema.Field(idx)
+		if field.Type.ID() == arrow.FixedWidthTypes.Timestamp_s.ID() || field.Type.ID() == arrow.FixedWidthTypes.Date32.ID() {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/test/go_integration_test_utils.go]: Yeah, there're `Timestamp_ms` and Date32 types that are not convertible to proto. We write Timestamp_s to parquet but somehow it's being read by Go parquet reader as `Timestamp_ms`. Reasons are unclear yet. Date32 is used system column `log_date`. We can skip it for now in tests.

File: go/internal/feast/logging/filelogsink.go (hunk 3 above)
Comment (pyalex) [reply go/internal/feast/logging/filelogsink.go]: fixed

File: go/internal/feast/logging/logger.go (hunk 4 above)
Comment (pyalex) [reply go/internal/feast/logging/logger.go]: Added comments to LogSink interface

File: go/internal/feast/logging/logger.go (hunk 5 above)
Comment (pyalex) [reply go/internal/feast/logging/logger.go:152]: Added comments to LogSink interface

File: go/internal/feast/logging/logger.go (hunk 4 above)
Comment (pyalex) [reply go/internal/feast/logging/logger.go]: We cannot really handle this error. But added printing

File: sdk/python/feast/feature_logging.py [hunk 11]
```diff
@@ -143,9 +143,11 @@ def to_data_source(self) -> DataSource:
 
 class LoggingConfig:
     destination: LoggingDestination
+    sample_rate: float
 
-    def __init__(self, destination: LoggingDestination):
+    def __init__(self, destination: LoggingDestination, sample_rate: float = 1.0):
```
[the remark this answers is not in the document]
Comment (pyalex) [reply sdk/python/feast/feature_logging.py:148]: The default behaviour is to omission of LoggingConfig in feature service. In this case no logging happening. But if you added logging config than default sample_rate is 100%.

File: go/internal/feast/model/featureservice.go [hunk 12]
```diff
@@ -11,17 +13,31 @@ type FeatureService struct {
 	CreatedTimestamp     *timestamppb.Timestamp
 	LastUpdatedTimestamp *timestamppb.Timestamp
 	Projections          []*FeatureViewProjection
+	LoggingConfig        *FeatureServiceLoggingConfig
+}
+
+type FeatureServiceLoggingConfig struct {
+	SampleRate        float32
+	PartitionInterval *durationpb.Duration
```
[the remark this answers is not in the document]
Comment (pyalex) [reply go/internal/feast/model/featureservice.go]: Not used right now. Deleted

**Review by kevjumba — APPROVED** [review]

**Pull request merged**